lib/AuthStateManager.ts

```diff
@@ -0,0 +1,155 @@
+    if (!sdk.emitter) {
+      throw new AuthSdkError('Emitter should be initialized before AuthStateManager');
+    }
+
+    this._sdk = sdk;
+    this._pending = { ...DEFAULT_PENDING };
+    this._authState = { ...DEFAULT_AUTH_STATE };
+
+    // Listen on tokenManager events to start updateState process
+    // "added" event is emitted in both add and renew process, it will be canceled by the followed "renewed" event
+    sdk.tokenManager.on('added', (key, token) => {
+      this.updateAuthState({ event: 'added', key, token });
+    });
+    sdk.tokenManager.on('renewed', (key, token) => {
+      this.updateAuthState({ event: 'renewed', key, token });
+    });
+    sdk.tokenManager.on('removed', (key, token) => {
+      this.updateAuthState({ event: 'removed', key, token });
+    });
+  }
+
+  getAuthState(): AuthState {
+    return this._authState;
+  }
+
+  updateAuthState({ event, key, token }: UpdateAuthStateOptions = {}): void {
+    const { isAuthenticated, devMode } = this._sdk.options;
+
+    const logger = (status) => {
+      console.group(`OKTA-AUTH-JS:updateAuthState: Event:${event} Status:${status}`);
+      console.log(key, token);
+      console.log('Current authState', this._authState);
+      console.groupEnd();
+    };
+
+    const emitAuthStateChange = (authState) => {
+      this._authState = authState;
+      // emit new authState object
+      this._sdk.emitter.emit(EVENT_AUTH_STATE_CHANGE, { ...authState });
+      devMode && logger('emitted');
+    };
+
+    if (this._pending.updateAuthStatePromise) {
+      if (this._pending.canceledTimes >= MAX_PROMISE_CANCEL_TIMES) {
+        // stop canceling then starting a new promise
+        // let existing promise finish to prevent running into loops
+        devMode && logger('terminated');
+        return;
+      } else {
+        this._pending.updateAuthStatePromise.cancel();
+      }
+    }
+
+    /* eslint-disable complexity */
+    const cancelablePromise = new PCancelable((resolve, _, onCancel) => {
+      onCancel.shouldReject = false;
+      onCancel(() => {
+        this._pending.updateAuthStatePromise = null;
+        this._pending.canceledTimes = this._pending.canceledTimes + 1;
+        devMode && logger('canceled');
+      });
+
+      const emitAndResolve = (authState) => {
+        if (cancelablePromise.isCanceled) {
+          resolve();
+          return;
+        }
+        // emit event and clear states
+        emitAuthStateChange(authState); 
+        this._pending = { ...DEFAULT_PENDING };
+        resolve();
+      };
+
+      this._sdk.tokenManager._getTokens()
+        .then(({ accessToken, idToken }) => {
+          if (cancelablePromise.isCanceled) {
+            resolve();
+            return;
+          }
+
+          // evaluate isPending to true if any token is expired
+          // then wait for next renewed event to evaluate a new state with valid tokens
+          // isPending state should only apply to token driven evaluation
```

Comment:
I think it's in a good state now.
For the initial flow, if there is one or more expired tokens in storage, set isPending to true, then wait for next evaluation to set the authState.
Also, check both tokens expiration state, then remove the expired one when the renew process failed. This can prevent authState from stucking on `isPending` state when both tokens are logically expired but only one get removed from storage. 
https://github.com/okta/okta-auth-js/pull/453/files#diff-f4d380d895ef82fba53a76eeed9089e1R221

Reply:
I think it can only be asssumed `isPending` if options.autoRenew is true, otherwise the expired tokens will stay expired and there will be no subsequent update.